core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala

```diff
@@ -106,18 +107,25 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
     val taCtxt = getTaskContext()
     val cmtr = getOutputCommitter()
     if (cmtr.needsTaskCommit(taCtxt)) {
-      try {
-        cmtr.commitTask(taCtxt)
-        logInfo (taID + ": Committed")
-      } catch {
-        case e: IOException => {
-          logError("Error committing the output of task: " + taID.value, e)
-          cmtr.abortTask(taCtxt)
-          throw e
+      val outputCommitCoordinator = SparkEnv.get.outputCommitCoordinator
+      val conf = SparkEnv.get.conf
+      val canCommit: Boolean = outputCommitCoordinator.canCommit(jobID, splitID, attemptID)
+      if (canCommit) {
```

Comment:
Makes sense.

I'm trying to implement this change now but am getting tripped up in unit tests again. Now, the unit tests need to capture the logic all the way down to the Executor. It looks like this workflow requires something closer to an end-to-end test now. Are there any unit tests that test scenarios with speculation? How should my existing OutputCommitCoordinator suite be adjusted?